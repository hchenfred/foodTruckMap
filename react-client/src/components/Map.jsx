import React, {Component} from 'react';
import foodImg from '../img/food.png';

class Map extends Component {
  constructor(props) {
    super(props);
    this.map = null;
    this.infowindow = null;
    this.createMarker = this.createMarker.bind(this);
    this.createAllMarkers = this.createAllMarkers.bind(this);
    this.clearMarkers = this.clearMarkers.bind(this);
  }

  componentDidMount() {
    this.initMap(this.props.currLocation);
  }

  initMap(currLocation) {
    this.map = new google.maps.Map(this.refs.map, {
      center: currLocation,
      zoom: 15
    });

    this.infowindow = new google.maps.InfoWindow();
    this.service = new google.maps.places.PlacesService(this.map);
    this.markers = [];
    fetch(`https://data.sfgov.org/resource/6a9r-agq8.json?$where=within_circle(location, 37.7749, -122.431297, 1000)&facilitytype='Truck'`)
    .then(response => response.json())
    .then(responseJson => {
      this.createAllMarkers(responseJson);
    });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps !== null && (nextProps.currLocation.lat !== this.props.currLocation.lat || nextProps.currLocation.lng !== this.props.currLocation.lng)) {
      this.clearMarkers();
      this.map.setCenter(nextProps.currLocation);
      fetch(`https://data.sfgov.org/resource/6a9r-agq8.json?$where=within_circle(location, ${nextProps.currLocation.lat}, ${nextProps.currLocation.lng}, 500)&facilitytype='Truck'`)
      .then(response => response.json())
      .then(responseJson => {
        console.log('hi', responseJson);
        this.createAllMarkers(responseJson);
      });
    }
  }

  clearMarkers() {
    this.markers.forEach(marker => {
      marker.setMap(null);
    })
  }

  createAllMarkers(trucks) {
    let tempResults = trucks;
    for (let i = 0; i < tempResults.length; i++) {
      let marker = this.createMarker(tempResults[i]);
      this.markers.push(marker);
      tempResults[i].marker = marker;
    }
    this.props.saveSearchedResults(tempResults);
  }

  createMarker(place) {
   const pinIcon = new google.maps.MarkerImage(
        foodImg,
        null, /* size is determined at runtime */
        null, /* origin is 0,0 */
        null, /* anchor is bottom center of the scaled image */
        new google.maps.Size(40, 40),
    );    
    const placeLoc = {lat: Number(place.latitude), lng: Number(place.longitude)};
    const marker = new google.maps.Marker({
      map: this.map,
      icon: pinIcon,
      position: placeLoc
    });

    const infowindow = new google.maps.InfoWindow({
          content: `<div>${place.applicant}</div>
          <div>${place.address}</div>`,
    });

    google.maps.event.addListener(marker, 'click', () => {
      infowindow.open(this.map, marker);
      setTimeout(() => { infowindow.close(); }, '1000');
    });

    return marker;
  }

  render() {
    return (
      <div ref="map" className="map" />
    );
  }
};

export default Map;