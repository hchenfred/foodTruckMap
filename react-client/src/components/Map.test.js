import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Map from './Map';

test('Map component should render as expected', () => {
  const component = shallow(<Map />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});