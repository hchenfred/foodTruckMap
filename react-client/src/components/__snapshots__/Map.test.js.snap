// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component should render as expected 1`] = `
<div
  className="map"
/>
`;
